### PR TITLE
Behavior metrics

### PR DESCRIPTION
**Pull request by pavi-rajes**
I added two functions:  time_to_target() and path_length() in analysis.py 
time_to_target() calculates reach time provided event codes and event times with an option to get per target statistics. Path length calculates the total distance traveled from cursor x and cursor y positions changing across time.  

In the same pull request, also adding another function get_trail_segments_and_times() in preproc.base. This function returns all corresponding event times along with event codes. 

Notes: 
Currently, time_to_target() only estimates reach times on successful targets. In the future, we can make it optional by adding an optional lamba reward filter function in the input. But I will treat this as an enhancement and this pull request need not wait for it.

### CODE REVIEW

File: aopy/analysis.py [hunk 1]
```diff
@@ -1543,3 +1543,61 @@ def interp_multichannel(x):
     x[nan_idx] = np.interp(idx,xp,fp)
 
     return x
+
+
+'''
+Behavioral metrics 
+'''
+
+def compute_path_length_per_trial(trajectory):
```
Comment (leoscholl) [inline aopy/analysis.py:1552]: why is it per trial? doesn't this just take a single trajectory

File: aopy/analysis.py [hunk 2]
```diff
@@ -1543,3 +1543,61 @@ def interp_multichannel(x):
...
+
+
+def time_to_target(event_codes, event_times, per_target_stats=False):
+    '''
+    This function calculates reach time to target only on rewarded trials given event codes and event times.
+
+    Args:
+        event_codes (list) : event codes
+        event_times (list) : event times corresponding to the event codes. These event codes and event times could be the output of preproc.base.get_trial_segments_and_times().
+         get_stats_per_target (bool): optional, use if you want to calculate reach time per target
```
Comment (leoscholl) [inline aopy/analysis.py:1574]: docstring arg doesn't match the input variable

File: aopy/analysis.py [hunk 3]
```diff
@@ -1543,3 +1543,61 @@ def interp_multichannel(x):
...
+
+    Returns:
+        reach_times (list):
+        reach_times_per_target(list of lists) : optional if per_target_stats == True
+    '''
+    CURSOR_ENTER_PERIPHERAL_TARGET = list(range(81, 89))
+    CENTER_TARGET_OFF = 32
+    REWARD = 48
```
Comment (leoscholl) [inline aopy/analysis.py:1582]: these should be inputs to the function in case they change for other experiments

File: aopy/analysis.py [hunk 4]
```diff
@@ -1543,3 +1543,61 @@ def interp_multichannel(x):
...
+    tr_e = np.array([event_codes[iTr] for iTr in range(len(event_times)) if REWARD in event_codes[iTr]])
+    tr_t = np.array([event_times[iTr] for iTr in range(len(event_times)) if REWARD in event_codes[iTr]])
```
Comment (leoscholl) [inline aopy/analysis.py:1584]: why is it necessary to restrict to rewarded trials? maybe leave this up to the user?

File: aopy/analysis.py [hunk 5]
```diff
@@ -1543,3 +1543,61 @@ def interp_multichannel(x):
...
+    return path_length
+
+
+def time_to_target(event_codes, event_times, per_target_stats=False):
+    '''
+    This function calculates reach time to target only on rewarded trials given event codes and event times.
+
+    Args:
+        event_codes (list) : event codes
+        event_times (list) : event times corresponding to the event codes. These event codes and event times could be the output of preproc.base.get_trial_segments_and_times().
```
Comment (leoscholl) [inline aopy/analysis.py:1573]: are these lists of events or lists of lists of events, where each list is a trial? in other words, is the input to this function supposed to be trial separated already?

File: aopy/analysis.py [hunk 6]
```diff
@@ -1543,3 +1543,61 @@ def interp_multichannel(x):
...
+    CURSOR_ENTER_PERIPHERAL_TARGET = list(range(81, 89))
+    CENTER_TARGET_OFF = 32
+    REWARD = 48
+    tr_e = np.array([event_codes[iTr] for iTr in range(len(event_times)) if REWARD in event_codes[iTr]])
+    tr_t = np.array([event_times[iTr] for iTr in range(len(event_times)) if REWARD in event_codes[iTr]])
+    leave_center_idx = np.argwhere(tr_e == CENTER_TARGET_OFF)[0, 1]
+
+    reach_target_idx = np.argwhere(np.isin(tr_e[0], CURSOR_ENTER_PERIPHERAL_TARGET))[0][0]
+
+    rt = tr_t[:, reach_target_idx] - tr_t[:, leave_center_idx]
```
Comment (leoscholl) [inline aopy/analysis.py:1589]: could you use slightly longer variable names -- all the rt / tr_t / tr_e is making it a bit hard to read

**Review by leoscholl — CHANGES_REQUESTED** [review]
some comments to address but overall looks like it works!

**Comment by pavi-rajes** [comment]
> some comments to address but overall looks like it works!

:) surely it works!! 

As I mentioned in the initial pull request, I will treat generalizing the reach time to include any trials as an enhancement to work in a separate pull request, since the function will break if the trial aligned event codes matrix is not of same length. We could include some checks or a condition that it should be a n_trials * 5 matrix, for example. I will fix your other comments on variable names. Thanks for reviewing!

File: aopy/analysis.py (hunk 1 above)
Comment (pavi-rajes) [reply aopy/analysis.py:1552]: Each trajectory is a cursor trajectory per trial. This function calculates path length per trial instead of path length for all trials in a session. Hence, the name says _per_trial

File: aopy/analysis.py (hunk 2 above)
Comment (pavi-rajes) [reply aopy/analysis.py]: fixed.

File: aopy/analysis.py (hunk 3 above)
Comment (pavi-rajes) [reply aopy/analysis.py]: fixed in a new commit

File: aopy/analysis.py (hunk 6 above)
Comment (pavi-rajes) [reply aopy/analysis.py]: fixed

File: aopy/analysis.py (hunk 5 above)
Comment (pavi-rajes) [reply aopy/analysis.py]: Yes, input to this function is trail-aligned event codes. Since we have another function to trial align event codes and time. It could be done outside the function.

**Comment by pavi-rajes** [comment]
Hold on on review. Some tests failed. I will fix it tomorrow.

File: aopy/analysis.py (hunk 1 above)
Comment (leoscholl) [reply aopy/analysis.py:1552]: i would call this one "calc_path_length" and the one that does all trials in a session "calc_path_length_per_trial". to me it makes more sense to have the "_per_trial" suffix on a function that cares about trials. this one you just give it a path and it computes the length, it doesn't need to know anything about trial structure.

File: aopy/analysis.py (hunk 5 above)
Comment (leoscholl) [reply aopy/analysis.py]: ok, can you please make it more clear in the docstring what the inputs are supposed to be? or give an example in the docstring?

File: aopy/analysis.py (hunk 1 above)
Comment (pavi-rajes) [reply aopy/analysis.py:1552]: I think _per_trial makes it explicit, and the function explanation says it is for a single trajectory. Other function names like time_to_target() or get_trial_segments() without _per_trial suffix are all used for batch processing. If that makes sense. I want to leave it as it is. If this causes further confusion, we can change it in a future patch.

File: aopy/analysis.py (hunk 5 above)
Comment (pavi-rajes) [reply aopy/analysis.py]: Already updated function documentation in a commit yesterday to make it clear. Please check latest version and see if further explanation is required.

File: aopy/analysis.py (hunk 1 above)
Comment (pavi-rajes) [reply aopy/analysis.py:1552]: It could be '_per_trajectory' insted of 'per_trial' perhaps. But either way i think a suffix if important.

**Review by leoscholl — APPROVED** [review]

File: aopy/analysis.py (hunk 1 above)
Comment (pavi-rajes) [reply aopy/analysis.py:1552]: haha. I will update the code to reflect this.

File: aopy/analysis.py [hunk 7]
```diff
@@ -1543,3 +1543,63 @@ def interp_multichannel(x):
...
+        only_rewarded_trials (bool, optional): only include trials with a reward code
+        per_target_stats (bool, optional): use if you want to calculate reach time per target
+
+    Returns:
+        reach_times (list):
+        reach_times_per_target(list of lists) : optional if per_target_stats == True
+    '''
+    event_times = np.array(event_times)
+    event_codes = np.array(event_codes)
+    if only_rewarded_trials:
```
Comment (pavi-rajes) [inline aopy/analysis.py:1586]: I am not sure I understand why this is helpful. The rest of the calculation outside the if loop will break if event_times and event_codes are a list of lists that are not of the uniform shape.

Comment (pavi-rajes) [reply aopy/analysis.py:1586]: One way of including a more generic trial filter is to get it as a lambda function instead.

Comment (pavi-rajes) [reply aopy/analysis.py:1586]: To be more clear, the function breaks because of some unsuccessful trials, such as the ones with center hold penalty with have shorter event_codes and event_times length, in which case, the rest of the calculations won't work. the condition for time to target is basically that the cursor enters the peripheral target.
This is what the lambda function could allow the user to choose. I hope my point gets across.

Comment (leoscholl) [reply aopy/analysis.py:1586]: hmm i see what you mean.  is it always going to be the case that rewarded trials are the same length? maybe we should just compute the lengths of the trials one at a time to avoid this issue in all cases?

Comment (pavi-rajes) [reply aopy/analysis.py:1586]: yes, always rewarded trials will have same event code sequence and will be of same length. I think we can keep the fucntion as it was before and add a note saying this will only work for rewarded trials. 
I will soon be generalizing to unrewarded trials with full trajectories as well. Like I said before, I will treat that as an enhancement.

Comment (leoscholl) [reply aopy/analysis.py]: i added another function called `calc_segment_duration` which does similar calculations but on any segment. i also found a bug in your code if there are more than 1 of the same target then the per target stats are wrong, they only take the first appearance of each target into account. i think you need to take a mean somewhere.

Comment (pavi-rajes) [reply aopy/analysis.py]: Could you explain a use case or the need for another function here? The time_to_target() does what it says it does. It would not be possible for me to test edge case scenarios extensively. However, I know this function works with our beignet's data and is generic enough to be applied to any session or task. 
The event codes for cursor entering peripheral targets are supposed to be unique. This function works as long as the codes are unique. Taking means within function would change the statistics when we concatenate reach times for example, so I avoided that.
